frame/staking/src/pallet/mod.rs

```diff
@@ -559,6 +651,184 @@ pub mod pallet {
+			ledger: &StakingLedger<T>,
+			validator: &T::AccountId,
+			page: PageIndex,
+		) -> bool {
+			ledger.legacy_claimed_rewards.binary_search(&era).is_ok() ||
+				Self::is_rewards_claimed(era, validator, page)
+		}
+
+		/// Check if the rewards for the given era and page index have been claimed.
+		///
+		/// This is only used for paged rewards. Once older non-paged rewards are no longer
+		/// relevant, `is_rewards_claimed_temp` can be removed and this function can be made public.
+		fn is_rewards_claimed(era: EraIndex, validator: &T::AccountId, page: PageIndex) -> bool {
+			ClaimedRewards::<T>::get(era, validator).binary_search(&page).is_ok()
+		}
+
+		/// Get exposure info for a validator at a given era and page.
+		///
+		/// This builds a paged exposure from `ExposureOverview` and `ExposurePage` of the
+		/// validator. For older non-paged exposure, it returns the clipped exposure directly.
+		pub(crate) fn get_validator_exposure(
+			era: EraIndex,
+			validator: &T::AccountId,
+			page: PageIndex,
+		) -> Option<ExposureExt<T::AccountId, BalanceOf<T>>> {
+			return match <ErasStakersPaged<T>>::get(era, (validator, page)) {
+				// return clipped exposure if page zero and paged exposure does not exist
+				None if page == 0 =>
+					Some(ExposureExt::from_clipped(<ErasStakersClipped<T>>::get(era, validator))),
+
+				// return paged exposure if it exists
+				Some(exposure_page) => {
+					let overview = <ErasStakersOverview<T>>::get(&era, validator);
+					// own stake is included only once in the first page.
+					let own = if page == 0 { overview.own } else { Zero::zero() };
+
+					Some(ExposureExt {
+						exposure_overview: ExposureOverview { own, ..overview },
+						exposure_page,
+					})
+				},
+				_ => None,
+			}
+		}
+
+		/// Returns the number of pages of exposure a validator has for the given era.
+		///
+		/// This will always return at minimum one count of exposure to be backward compatible to
+		/// non-paged reward payouts.
+		pub(crate) fn get_page_count(era: EraIndex, validator: &T::AccountId) -> PageIndex {
+			<ErasStakersOverview<T>>::get(&era, validator).page_count.max(1)
+		}
+
+		/// Returns the next page that can be claimed or `None` if nothing to claim.
+		pub(crate) fn get_next_claimable_page(
+			era: EraIndex,
+			validator: &T::AccountId,
+			ledger: &StakingLedger<T>,
+		) -> Option<PageIndex> {
+			if Self::is_non_paged_exposure(era, validator) {
+				return match ledger.legacy_claimed_rewards.binary_search(&era) {
+					// already claimed
+					Ok(_) => None,
+					// Non-paged exposure is considered as a single page
+					Err(_) => Some(0),
+				}
+			}
+
+			// Find next claimable page of paged exposure.
+			let page_count = Self::get_page_count(era, validator);
+			let claimed_pages = ClaimedRewards::<T>::get(era, validator);
+			let claimed_page_count = claimed_pages.len() as PageIndex;
+
+			// find the first page that is not claimed.
+			for page in 0..claimed_page_count as PageIndex {
```

Comment:
`claimed_page_count` is already `PageIndex`.

Reply:
I think for just a few iterations, I think code does not need to be very efficient and can perhaps be written a bit simpler. Here's my suggestion: 

1. get `claimed_pages`.
2. construct `all_claimable_pages = (0..page_count).collect()`.
3. the first item that is in `all_claimable_pages` but not in `claimed_pages` is our target, which means: `next_claimable = all_claimable_pages.iter().filter(|x| !claimed_pages.contains(x)).next()`. This gives you a nice `Option` that you can use actually.